src/Compilers/CSharp/Portable/Symbols/Source/QuickAttributeChecker.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// The QuickAttributeChecker applies a simple fast heuristic for determining probable
+    /// attributes of certain kinds without binding attribute types, just by looking at the final syntax of an
+    /// attribute usage.
+    /// </summary>
+    /// <remarks>
+    /// It works by maintaining a dictionary of all possible simple names that might map to the given
+    /// attribute.
+    /// </remarks>
+    internal sealed class QuickAttributeChecker
+    {
+        private readonly Dictionary<string, QuickAttributes> _nameToAttributeMap;
+        private static QuickAttributeChecker _lazyQuickAttributeChecker;
+
+#if DEBUG
+        private bool _sealed;
+#endif
+
+        internal static QuickAttributeChecker Predefined
+        {
+            get
+            {
+                if (_lazyQuickAttributeChecker is null)
+                {
+                    Interlocked.CompareExchange(ref _lazyQuickAttributeChecker, CreateQuickAttributeChecker(), null);
+                }
+
+                return _lazyQuickAttributeChecker;
+            }
+        }
+
+        private static QuickAttributeChecker CreateQuickAttributeChecker()
+        {
+            var result = new QuickAttributeChecker();
+            result.AddName(AttributeDescription.TypeIdentifierAttribute.Name, QuickAttributes.TypeIdentifier);
+            result.AddName(AttributeDescription.TypeForwardedToAttribute.Name, QuickAttributes.TypeForwardedTo);
+
+#if DEBUG
+            result._sealed = true;
+#endif
+            return result;
+        }
+
+        private QuickAttributeChecker()
+        {
+            _nameToAttributeMap = new Dictionary<string, QuickAttributes>();
+            // NOTE: caller must seal
+        }
+
+        private QuickAttributeChecker(QuickAttributeChecker previous)
+        {
+            _nameToAttributeMap = new Dictionary<string, QuickAttributes>(previous._nameToAttributeMap);
+            // NOTE: caller must seal
+        }
+
+        private void AddName(string name, QuickAttributes newAttributes)
+        {
+#if DEBUG
+            Debug.Assert(!_sealed);
+#endif
+            var currentValue = QuickAttributes.None;
+            _nameToAttributeMap.TryGetValue(name, out currentValue);
+
+            QuickAttributes newValue = newAttributes | currentValue;
```

Comment:
>QuickAttributes newValue = newAttributes | currentValue; [](start = 12, length = 56)

It feels like we should have targeted tests for the merging logic here.

Reply:
Would direct test on QuickAttributeChecker be ok?
It seems difficult to observe this at compilation level.

---
In reply to: [155960702](https://github.com/dotnet/roslyn/pull/23647#discussion_r155960702) [](ancestors = 155960702)

Reply:
> Would direct test on QuickAttributeChecker be ok?

I think it would be Ok.

---
In reply to: [156202384](https://github.com/dotnet/roslyn/pull/23647#discussion_r156202384) [](ancestors = 156202384,155960702)